zerver/lib/user_groups.py

```diff
@@ -524,27 +523,58 @@ def get_group_setting_value_for_api(
     )
 
 
+@dataclass
+class UserGroupMembersDict:
+    direct_members: list[int]
+    direct_subgroups: list[int]
+
+
 def get_setting_value_for_user_group_object(
-    setting_value_group: UserGroup,
-    direct_members_dict: dict[int, list[int]],
-    direct_subgroups_dict: dict[int, list[int]],
+    setting_group_id: int,
+    named_user_group_ids: set[int],
+    members_dict: dict[int, UserGroupMembersDict],
 ) -> int | AnonymousSettingGroupDict:
-    if hasattr(setting_value_group, "named_user_group"):
-        return setting_value_group.id
+    if setting_group_id in named_user_group_ids:
+        return setting_group_id
 
-    direct_members = []
-    if setting_value_group.id in direct_members_dict:
-        direct_members = direct_members_dict[setting_value_group.id]
+    return AnonymousSettingGroupDict(
+        direct_members=members_dict[setting_group_id].direct_members,
+        direct_subgroups=members_dict[setting_group_id].direct_subgroups,
+    )
 
-    direct_subgroups = []
-    if setting_value_group.id in direct_subgroups_dict:
-        direct_subgroups = direct_subgroups_dict[setting_value_group.id]
 
-    return AnonymousSettingGroupDict(
-        direct_members=direct_members,
-        direct_subgroups=direct_subgroups,
+def get_members_and_subgroups_of_groups(group_ids: list[int]) -> dict[int, UserGroupMembersDict]:
+    user_members = (
+        UserGroupMembership.objects.filter(user_group_id__in=group_ids)
+        .exclude(user_profile__is_active=False)
```

Comment:
Same queries is used for fetching members for stream group settings as well, except that there deactivated users are not excluded, which should be excluded. Then we started work on #33371, so it is pending to be fixed. Maybe we can fi that first, which will deduplicate the code as well and then complete #33371.

Reply:
Yeah, we may be able to combine these nicely.

Reply:
I was thinking of first fixing the stream queries to not included deactivated users in a separate PR and that would do deduplication as well. And then we can see when can we complete #33371.

Reply:
Ahh, yes that sounds reasonable.